kill/kill.1
.\"
.\" Sccsid @(#)kill.1	1.2 (gritter) 6/30/05
.\" Derived from kill(1), Unix 7th edition:
.\" Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
.\"
.\" SPDX-Licence-Identifier: Caldera
.\"
.TH KILL 1 "6/30/05" "Heirloom Toolchest" "User Commands"
.SH NAME
kill \- terminate a process
.SH SYNOPSIS
\fBkill\fR [\fB\-\fIsignal\fR|\fB\-s\ \fIsignal\fR] \fIprocessid\fR ...
.br
\fBkill\fR \fB\-l\fR [\fIstatus\fR]
.SH DESCRIPTION
.I Kill
sends signal 15 (terminate) to the specified processes.
If a signal number or symbolic signal name preceded by `\-' or
.B \-s
is given as first argument, that signal is sent instead of
terminate
(see
.IR  signal (2)).
This will kill processes that do not catch the signal;
in particular `kill \-9 ...' is a sure kill.
.PP
By convention, if process number 0 is specified, all members
in the process group (i.e. processes resulting from
the current login) are signaled.
If process number \-1 is specified,
all processes for which the user is permitted to send a signal are signaled.
Otherwise if the process number is negative,
it is treated as a process group id,
and all processes in that group are signaled.
.PP
With the
.B \-l
option, all available symbolic signal names are listed.
If the
.I status
argument is present,
the symbolic signal name corresponding to
.IR status ,
interpreted either as a signal number or as an exit value
as given by the `$?' shell parameter,
is printed.
.PP
A symbolic signal name consists of the name of the signal
in the `C' language without the `SIG' prefix,
e.\|g. `QUIT' for `SIGQUIT'.
.PP
The killed processes must belong
to the current user unless
he is the super-user.
.PP
The process number of an asynchronous process
started with `&' is reported by the shell.
Process numbers can also be found by using
.IR ps (1).
.SH "SEE ALSO"
ps(1), kill(2), signal(2)


CHANGES
Release 250220
* seq, write: Fix memory fault due to illegal pointer access.
* cpio, pax: Fix manual page formatting.

Release 250209
* cpio: Fix manual page formatting.
* apply: Fixed the command string magic character computation algorithm that
  got garbled after a refactor made in the name of performance.
* build/mk.config, chroot, nawk: Adding -ltinfo for static compilation with
  glibc and making changes to adhere to C standards (patch by Gabriel Souza).
* [NEW] wall, write: New implementation ported from UNIX 7th (a.k.a. 32V),
  bulletproofed against buffer overflows and the CVE-2024-28085, which is
  available as the 'ssafe()' function at libcommon.
* rm: Alias '-e' flag to more known '-v' for verbose.
* date, shl: Fix compilation on platforms where (struct utmpx).ut_tv is not
  of type 'timeval'.
* rev: UTF-8 support, borrowed from suckless sbase. In varietate concordia.
* [NEW] apply, pick: Complete implementation for both apply and pick, from
  Research UNIX 8th and 10th respectively; extended payloads for pick in
  comparison with the original implementation.
* seq: Support for counting from negative numbers to positive or making a
  negative sequence.
* rev: Full refactor (made by Samuel Brederodes).
* Fixes for building dynamically on Arch Linux and other GNU C
  Library-based Linux systems which ships exclusively dynamic versions
  of its libraries, also remove `readlink_ucb' object file at `readlink/'
  build directory when executing `make clean' (patch by Molly A. McCollum).
* Remove `register's from legacy code that was living unchanged since at
  least the 80's, since modern linters started complaining about it and
  it is not used as much on most common architectures such as x86
  (patch by Samuel Brederodes).
* watch: `use_default_colors()', do not force black and white.
* [NEW] rev: New implementation, not capable of UTF-8 yet.
* cpio, diff, tabs: Fix duplicated definitions (made/reported by crumbtoo,
  merged by me, Luiz Antônio Rangel). Thanks! --- or better: Merci!
* rm: Added real support for the POSIX 1003.1(2008) standard, implemented
  IBM AIX(R)-style verbosing and 4.4BSD's '-d' option for directories ---
  rmdir(1)-esque payload --- as extensions to the standard.
* seq: Made its payload similar to UNIX v8's implementation, but with some
  extensions such as a fast algorithm for -w or -s --- which was already
  present before the (almost) full rewrite.
* timeout: avoid buffer overrun, also did some changes at the source code,
  such as cleaning up comments and initializing 'commandv' and 'fst_commandv'.
* watch: Used esr's deheader for removing unnecessary headers, also use
  pfmt(stderr, MM_ERROR, ...)  instead of prerror().
* watch, timeout: Some small corrections at the manual page.
* cpio: Skip over zip extra headers with fewer than four bytes (the minimum
  length for them to contain any information) instead of reporting a bad
  header and terminating.
* pg: Handle cases where standard output is a terminal but not open for
  reading (bug reported by Keith Doyle).
* cpio, tar: If an archive contains a file that is a hardlink to itself on
  extraction, do nothing at all instead of discarding its data. This mainly
  happens if a file with more than one link is included under the same name
  in an archive twice. Since special situations on extraction such as bind
  mounts on Linux may lead to this scenario even in cases where both link
  names are different, this is now treated like a successful link() step
  (bugreport by Thomas Graf).
* Made the getopt() substitute work with the getopt$UNIX2003() function
  redirection mechanism on Mac OS X Leopard.
* mknod, pgrep, ps: Mac OS X Leopard compilation fixes.
* w, whodo: On a 32-bit system where /proc/kcore is larger than 2 GB,
  avoid a harmless error message (bug reported by Pierre Gentile).
* ptime: Works on recent Linux kernels again.
* col: The behavior of the -x flag has been inverted to actually match
  the documentation, i.e. to suppress conversion of spaces to tabs (patch
  by Markus Schnalke).
* cut: With "-s", print the last field of a line that was also the last
  field specified with "-f" (bugreport by Pierre Gentile).

Release 240220
* [NEW] seq: Stone-simple implementation; needs some improvement yet.
* [NEW] watch: Complete, simple and stable implementation.
* [NEW] timeout: Complete implementation with extended payloads.
* readlink: Different help messages for both UCB and default variants.

Release 230503
* chroot: "docs: Update date at chroot.8 manual page."
* chroot: "chore: Remove comment, will move to GitHub"
* chroot: "docs: Rewrote chroot(8) manual page"
* chroot: "chore: Add ``usage()`` function at chroot.c"
* tabs: "fix: [t,g]roff formatting at tabs/fspec.5."
* readlink: "docs: Created new manual page for the non-UCB variant. (#24)"
* Makefile.mk: "fix: Don't symbolic/hard link readlink from UCBDIR to DEFDIR anymore"
* readlink: "docs: Add comments clarifying the C preprocessor directives."
* readlink: "fix: Ignore ``-n`` option when operating multiple files."
* readlink: "feat: Create non-UCB/GNU-alike variant."
* readlink: "fix: Made ``name_size`` private again, so ``pfmt()`` won't go segmentation fault."
* readlink: "feat: First step, start modularizing the code."
* libcommon: "fix: getdirentries replacement for OBSD >5.4"
* fix-me: For new releases, finish porting for OpenBSD and implement watch(1) and timeout(1).

Release 230305
* getconf: "Now _NPROCESSORS_ONLN and _NPROCESSORS_CONF are implemented."
* (*): "Removed the remaining licence headers, now every licence is referenced by its SPDX identifier."

Release 230228
* [NEW] readlink: New implementation, extensively based on OpenBSD's.
* readlink: It's installed at /usr/ucb per default and symbolic/hard linked to /usr/5bin.

Release 230120
* Notes for the library directories if linking against the GNU C library.

kill/kill.c
/*
 * kill - terminate a process
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, June 2005.
 */
/*
 * Copyright (c) 2005 Gunnar Ritter
 *
 * SPDX-Licence-Identifier: Zlib
 */

/*
 * Sccsid @(#)kill.c	1.2 (gritter) 6/30/05
 */

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <errno.h>
#include <signal.h>
#include <libgen.h>
#include <ctype.h>
#include <errno.h>

#ifdef	SIGRTMIN
#undef	NSIG
#define	NSIG	SIGRTMIN
#endif
#ifndef	NSIG
#define	NSIG	32
#endif

static const char	*progname;
static int		lflag;
static int		sflag = SIGTERM;
static int		status;

extern int	sig_2_str(int, char *);
extern int	str_2_sig(const char *, int *);

static void
usage(void)
{
	fprintf(stderr, "usage: %s [ [ -sig ] id ... | -l ]\n", progname);
	exit(1);
}

static void
badsig(void)
{
	fprintf(stderr, "%s: bad signal\n", progname);
	exit(1);
}

static int
getsig(const char *cp)
{
	int	n;

	if (str_2_sig(cp, &n) < 0)
		badsig();
	return n;
}

static void
listsigs(void)
{
	int	i, j = 0;
	char	b[20];

	for (i = 1; i < NSIG; i++) {
		if (sig_2_str(i, b) < 0)
			continue;
		if (j++ % 10)
			putchar('\t');
		else if (j > 1)
			putchar('\n');
		fputs(b, stdout);
	}
	putchar('\n');
}

static void
printsig(const char *cp)
{
	char	*xp;
	int	n;
	char	b[20];

	n = strtol(cp, &xp, 10);
	if (*xp != '\0' || *cp == '-' || *cp == '+')
		badsig();
	if (n >= 128)
		n -= 128;
	if (sig_2_str(n, b) < 0)
		badsig();
	puts(b);
}

static void
sendsig(const char *pp)
{
	pid_t	pid;
	char	*xp;

	pid = strtol(pp, &xp, 10);
	if (*xp != '\0' || *pp == '+')
		goto esrch;
	if (kill(pid, sflag) < 0) {
		switch (errno) {
		case EINVAL:
			badsig();
			/*NOTREACHED*/
		case EPERM:
			fprintf(stderr, "%s: permission denied\n", progname);
			status |= 1;
			break;
		case ESRCH:
		esrch:	fprintf(stderr, "%s: %s\n", progname, pid >= 0 ?
				"no such process" : "no such process group");
			status |= 1;
			break;
		default:
			perror(progname);
			status |= 1;
		}
	}
}

int
main(int argc, char **argv)
{
	int	i = 0;

	progname = basename(argv[0]);
	if (argc >= 2 && argv[1][0] == '-') {
		i++;
		if (argv[1][1] == 's') {
			if (argv[1][2])
				sflag = getsig(&argv[1][2]);
			else if (argc >= 3) {
				i++;
				sflag = getsig(&argv[2][0]);
			} else
				usage();
		} else if (argv[1][1] == 'l')
			lflag = 1;
		else
			sflag = getsig(&argv[1][1]);
	}
	argv += i;
	argc -= i;
	if (lflag) {
		if (argc == 1)
			listsigs();
		else if (argc == 2)
			printsig(argv[1]);
		else
			usage();
	} else {
		if (argc <= 1)
			usage();
		for (i = 1; i < argc; i++)
			sendsig(argv[i]);
	}
	return status;
}
